packages/native-stack/src/views/RetainContext.tsx

```diff
@@ -0,0 +1,138 @@
+  retain(): string;
+  release(key: string): boolean;
+  restore(key: string): boolean;
+  supported: boolean;
+};
+
+/**
+ * This context allows you to retain screens in memory, even after
+ * they have been popped off the stack. This is important for
+ * supporting native iOS "Picture in Picture" mode, as the
+ * originating UIViewController must be retained in memory for the
+ * PiP overlay to continue running.
+ * Retained screens can also be "restored" to the top of the stack,
+ * for example when the user taps the button in the PiP overlay
+ * to return to the app.
+ */
+export const RetainContext = React.createContext<RetainContextT>({
+  retain() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  release() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  restore() {
+    throw new Error(`Not in a native screen stack`);
+  },
+  supported: false,
+});
+
+type RetainedScenes = {
+  [key: string]: {
+    route: Route<any>;
+    descriptor: NativeStackDescriptorMap['key'];
+  };
+};
+
+export function useRetainContext(
+  state: StackNavigationState<ParamListBase>,
+  navigation: NativeStackNavigationHelpers,
+  descriptors: NativeStackDescriptorMap
+) {
+  const [retainedScenes, setRetainedScenes] = React.useState<RetainedScenes>(
+    {}
+  );
+  const latestValues = React.useRef({ state, retainedScenes, descriptors });
+  React.useEffect(() => {
+    latestValues.current = { state, retainedScenes, descriptors };
+  });
+
+  const retainContext = React.useMemo<RetainContextT>(
+    () => ({
+      retain() {
+        const { state, descriptors } = latestValues.current;
+        const route = state.routes[state.routes.length - 1];
+        setRetainedScenes((screens) => ({
+          ...screens,
+          [route.key]: {
+            route,
+            descriptor: descriptors[route.key],
+          },
+        }));
+        return route.key;
+      },
+      release(key) {
+        setRetainedScenes((screens) => {
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          const { [key]: _, ...rest } = screens;
+          return rest;
+        });
+        return latestValues.current.retainedScenes[key] != null;
+      },
+      restore(key) {
+        const { retainedScenes, state } = latestValues.current;
+
+        const route = retainedScenes[key]?.route;
+        if (!route) {
+          return false;
+        }
+
+        // Remove from retained
+        setRetainedScenes((screens) => {
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          const { [key]: _, ...rest } = screens;
+          return rest;
+        });
+
+        const index = state.routes.findIndex((r) => r.key === key);
+        const routes =
+          index !== -1 ? state.routes.slice(0, index) : state.routes.slice();
+        routes.push(route);
+        navigation.dispatch({
+          type: 'RESET',
+          payload: { ...state, index: routes.length - 1, routes },
+        });
```

Comment:
Don't refer to actions by string. Use the `CommonActions.reset` action creator instead. In this case it should be possible to just use `navigation.reset`.

Also, I assume it's moving the screen to the end. Is this necessary/desirable? If the screen wasn't the last one, seems strange to change the order after pip restore. I'd remove this, consumers can do this themselves if they want to.

Reply:
> If the screen wasn't the last one, seems strange to change the order after pip restore.

I thought that in order for a screen to be visible on a screen stack, it needed to be the last one. Is that not the case? `restore()` is supposed to bring the screen back to visibility; either adding it to the end if it had been popped off, or popping off anything on top of it if the screen still exists in the stack.